snowpack/src/build/import-resolver.ts

```diff
@@ -44,7 +47,24 @@ function resolveSourceSpecifier(lazyFileLoc: string, config: SnowpackConfig) {
       lazyFileLoc = tsWorkaroundImportFileLoc;
     }
   } else {
-    lazyFileLoc = lazyFileLoc + '.js';
+    // missing extension
+    if (getFsStat(lazyFileLoc + path.extname(parentFile))) {
```

Comment:
This is the main change here. If an extension is missing:

1. First, it tries its parent extension (assume that `.ts` assumes `.ts`, `.scss` assumes `.scss`, `.svelte` assumes `.svelte`, etc.
2. If that fails, then blindly loop through all `config._extensionMap` types. Find the first match (in the event you have the same file in the same folder that differs by extension, we may grab the wrong one, but how were we supposed to know?)
3. If all that fails, then fall back to `.js`.

This PR should be an improvement over existing behavior, because before, we’d jump straight to #3 with no warning. While I think that’d be OK if we threw a warning, we don’t—we’d just silently assume `.js` even when there were better guesses.

Reply:
This is obviously a really clever feature! But... do we want to support this? Node.js ESM is going more strict, not less (this wouldn't be supported in Node) and I'm worried that allowing this will encourage users do do it across their project, and result in less efficient code. Not supporting this has been intentional up to this point (1 exception: TS, which literally won't let you import with the `.ts` extension).

Thoughts on moving this info into a warning instead, to use this info to help you rewrite your import correctly?

Reply:
In JS I’d be +1 for an error, but in a framework like Vue or Svelte, or a preprocessor like Sass, it just feels pedantic. The current behavior is broken—it just crashes without an actionable error—so current behavior does need to change. While I always want to follow ESM rules, I do somewhat feel like ESM rules don’t always apply in invented languages. And many invented languages omit extensions.

If we want to encourage people to add extensions and prepare people for ESM, I’d be open to a warning, but I’d really want it to compile properly with that warning. As this PR shows it’s an easy add for us, and it’s not like browsers will ever support `.vue` files anyway (well, not anytime soon), so we could at least just build.

Reply:
But also, this is a regression—a Vue site I had that built successfully on `3.0` fails to build in `3.1` without an explanation. So I do think this PR is blocking a `3.1` release.

Reply:
100% agreed that crashing is no good here, I'm not pitching that. But I really am having trouble feeling comfortable with this from a performance story (if we don't warn, then your whole app could be built to use these lazy imports) and from a stability story (undefined which file you mean, if you don't include it. An `index.ts` file importing an `App.tsx` but not an `App.css` file exists even in our simple starter apps, which would be undefined if left without an extension based on this logic).

Node.js is taking a lot of pains to push the "extensions are required" story with ESM, to much handwringing. It's as much a philosophical question (matching the browser where we can) as it is a feature set.

What else can we do about the regression itself? Is it that you had a `./App` import that should have resolved to `./App.vue` but not doesn't resolve anywhere? If so, lets add back any "`.js` fallback on missing" behavior that was removed in 3.1.

Reply:
> What else can we do about the regression itself? Is it that you had a `./App` import that should have resolved to `./App.vue` but not doesn't resolve anywhere?

No, in the PR notes above you’ll see that Snowpack’s resolution changed in `3.1`. In a `.vue` file I have an import for `import Component from './Component'`. In `3.0`, Snowpack would first try and resolve that, and it would find `./Component.vue`. In `3.1`, Snowpack first appends `.js`, and then tries to resolve that, and fails.

In the PR notes above, this adds 2 steps to the resolution logic: first, check parent extension; second, check extension map. WDYT about only doing the first step? That shouldn’t really affect the performance story too much (hopefully). And we can still add the “hey this file ____ is missing an extension, please fix” but we keep existing behavior from `3.0`.

Reply:
> Node.js is taking a lot of pains to push the "extensions are required" story with ESM, to much handwringing and complaints.

Agreed 100% about encouraging that. But where I’m scared we’d get pushback is when this overreaches into other third-party languages that have ways of doing things (often extensionless). I just don’t want someone to have a bad experience in Snowpack following a tutorial in the tool they’re using and getting different behavior only in Snowpack.

Reply:
Okay, appreciate the extra details. If we're just trying to match v3.0 behavior, then you're right that we should map through the extensionMap, and check for any files but only ones **that build to JS**. If it builds to anything else, that was never supported in v3.0 (since our fallback was always a ".js" fallback using the final URL)

We actually store a map of all files in memory, so maybe refactoring this to use that (in a separate PR) could get around my performance concerns. And then we can revisit this in an eventual v4.0, maybe with a warning.

So I guess the ask here is to filter the `_extensionMap` search below to only check files that build to JS? I can get behind that.

Reply:
Makes sense! I like that solution.

Reply:
OK updated. PTAL when you get another chance!